File: samples/Android/xtruntime/src/main/java/com/opensource/xtruntime/XTRCanvasView.kt
package com.opensource.xtruntime

import android.graphics.*
import com.eclipsesource.v8.V8
import com.eclipsesource.v8.V8Object
import org.mozilla.javascript.ScriptableObject

/**
 * Created by cuiminghui on 2017/9/22.
 */
class XTRCanvasView: XTRComponent() {

    override val name: String = "XTRCanvasView"

    override fun v8Object(): V8Object? {
        val v8Object = V8Object(xtrContext.v8Runtime)
        v8Object.registerJavaMethod(this, "createScriptObject", "createScriptObject", arrayOf(V8Object::class.java, V8Object::class.java))
        return v8Object
    }

    fun createScriptObject(rect: V8Object, scriptObject: V8Object): V8Object {
        val view = InnerObject(scriptObject.twin(), xtrContext)
        XTRUtils.toRect(rect)?.let {
            view.frame = it
        }
        return view.requestV8Object(xtrContext.v8Runtime)
    }

    class State(var globalAlpha: Double = 1.0,
                var fillStyle: XTRColor? = null,
                var strokeStyle: XTRColor? = null,
                var lineCap: String? = null,
                var lineJoin: String? = null,
                var lineWidth: Double = 1.0,
                var miterLimit: Double = 0.0,
                var currentTransform: Matrix = Matrix()) {
        fun copy(): State {
            return State(globalAlpha, fillStyle, strokeStyle, lineCap, lineJoin, lineWidth, miterLimit, currentTransform)
        }
    }

    class InnerObject(scriptObject: V8Object, xtrContext: XTRContext) : XTRView.InnerObject(scriptObject, xtrContext), XTRObject {

        private var actions: List<(canvas: Canvas) -> Unit> = listOf()
        private var currentState = State()
        private var fakeState = State()
        private var stateStack: List<State> = listOf()
        private var currentPath = Path()
        private var drawingPath = Path()
        private var currentPaint = Paint()
        private val scale = resources.displayMetrics.density


        override fun requestV8Object(runtime: V8): V8Object {
            val v8Object = super<XTRView.InnerObject>.requestV8Object(runtime)
            v8Object.registerJavaMethod(this, "xtr_globalAlpha", "xtr_globalAlpha", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setGlobalAlpha", "xtr_setGlobalAlpha", arrayOf(Double::class.java))
            v8Object.registerJavaMethod(this, "xtr_fillStyle", "xtr_fillStyle", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setFillStyle", "xtr_setFillStyle", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_strokeStyle", "xtr_strokeStyle", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setStrokeStyle", "xtr_setStrokeStyle", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_lineCap", "xtr_lineCap", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setLineCap", "xtr_setLineCap", arrayOf(String::class.java))
            v8Object.registerJavaMethod(this, "xtr_lineJoin", "xtr_lineJoin", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setLineJoin", "xtr_setLineJoin", arrayOf(String::class.java))
            v8Object.registerJavaMethod(this, "xtr_lineWidth", "xtr_lineWidth", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setLineWidth", "xtr_setLineWidth", arrayOf(Double::class.java))
            v8Object.registerJavaMethod(this, "xtr_miterLimit", "xtr_miterLimit", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_setMiterLimit", "xtr_setMiterLimit", arrayOf(Double::class.java))
            v8Object.registerJavaMethod(this, "xtr_rect", "xtr_rect", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_fillRect", "xtr_fillRect", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_strokeRect", "xtr_strokeRect", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_fill", "xtr_fill", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_stroke", "xtr_stroke", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_beginPath", "xtr_beginPath", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_moveTo", "xtr_moveTo", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_closePath", "xtr_closePath", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_lineTo", "xtr_lineTo", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_quadraticCurveTo", "xtr_quadraticCurveTo", arrayOf(V8Object::class.java, V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_bezierCurveTo", "xtr_bezierCurveTo", arrayOf(V8Object::class.java, V8Object::class.java, V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_arc", "xtr_arc", arrayOf(V8Object::class.java, Double::class.java, Double::class.java, Double::class.java, Boolean::class.java))
            v8Object.registerJavaMethod(this, "xtr_isPointInPath", "xtr_isPointInPath", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_postScale", "xtr_postScale", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_postRotate", "xtr_postRotate", arrayOf(Double::class.java))
            v8Object.registerJavaMethod(this, "xtr_postTranslate", "xtr_postTranslate", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_postTransform", "xtr_postTransform", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_setCanvasTransform", "xtr_setCanvasTransform", arrayOf(V8Object::class.java))
            v8Object.registerJavaMethod(this, "xtr_save", "xtr_save", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_restore", "xtr_restore", arrayOf())
            v8Object.registerJavaMethod(this, "xtr_clear", "xtr_clear", arrayOf())
            return v8Object
        }

        fun xtr_globalAlpha(): Double {
            return fakeState.globalAlpha
        }

        fun xtr_setGlobalAlpha(value: Double) {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentState.globalAlpha = value
                }
                this.actions = actions.toList()
            }
            fakeState.globalAlpha = value
        }

        fun xtr_fillStyle(): XTRColor? {
            return fakeState.fillStyle
        }

        fun xtr_setFillStyle(value: V8Object) {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentState.fillStyle = XTRUtils.toColor(value)
                }
                this.actions = actions.toList()
            }
            fakeState.fillStyle = XTRUtils.toColor(value)
        }

        fun xtr_strokeStyle(): XTRColor? {
            return fakeState.strokeStyle
        }

        fun xtr_setStrokeStyle(value: V8Object) {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentState.strokeStyle = XTRUtils.toColor(value)
                }
                this.actions = actions.toList()
            }
            fakeState.strokeStyle = XTRUtils.toColor(value)
        }

        fun xtr_lineCap(): String? {
            return fakeState.lineCap
        }

        fun xtr_setLineCap(value: String) {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentState.lineCap = value
                }
                this.actions = actions.toList()
            }
            fakeState.lineCap = value
        }

        fun xtr_lineJoin(): String? {
            return fakeState.lineJoin
        }

        fun xtr_setLineJoin(value: String) {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentState.lineJoin = value
                }
                this.actions = actions.toList()
            }
            fakeState.lineJoin = value
        }

        fun xtr_lineWidth(): Double {
            return fakeState.lineWidth
        }

        fun xtr_setLineWidth(value: Double) {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentState.lineWidth = value
                }
                this.actions = actions.toList()
            }
            fakeState.lineWidth = value
        }

        fun xtr_miterLimit(): Double {
            return fakeState.miterLimit
        }

        fun xtr_setMiterLimit(value: Double) {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentState.miterLimit = value
                }
                this.actions = actions.toList()
            }
            fakeState.miterLimit = value
        }

        fun xtr_rect(value: V8Object) {
            XTRUtils.toRect(value)?.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentPath.reset()
                        currentPath.addRect(
                                RectF(
                                        it.x.toFloat() * scale,
                                        it.y.toFloat() * scale,
                                        (it.x + it.width).toFloat() * scale,
                                        (it.y + it.height).toFloat() * scale
                                ), Path.Direction.CCW)
                    }
                    this.actions = actions.toList()
                }
            }
        }

        fun xtr_fillRect(value: V8Object) {
            xtr_rect(value)
            xtr_fill()
        }

        fun xtr_strokeRect(value: V8Object) {
            xtr_rect(value)
            xtr_stroke()
        }

        fun xtr_fill() {
            this.actions.toMutableList()?.let { actions ->
                actions.add { currentCanvas ->
                    currentPaint.reset()
                    currentPaint.color = (currentState.fillStyle ?: XTRColor(0.0, 0.0, 0.0, 1.0)).intColor()
                    currentPaint.alpha = (currentState.globalAlpha * 255.0).toInt()
                    if (!currentState.currentTransform.isIdentity) {
                        drawingPath.reset()
                        drawingPath.addPath(currentPath)
                        drawingPath.transform(currentState.currentTransform)
                        currentCanvas.drawPath(drawingPath, currentPaint)
                        return@add
                    }
                    currentCanvas.drawPath(currentPath, currentPaint)
                }
                this.actions = actions.toList()
            }
            invalidate()
        }

        fun xtr_stroke() {
            this.actions.toMutableList()?.let { actions ->
                actions.add { currentCanvas ->
                    currentPaint.reset()
                    currentPaint.color = (currentState.strokeStyle ?: XTRColor(0.0, 0.0, 0.0, 1.0)).intColor()
                    currentPaint.alpha = (currentState.globalAlpha * 255.0).toInt()
                    currentPaint.style = Paint.Style.STROKE
                    when (currentState.lineCap) {
                        "butt" -> currentPaint.strokeCap = Paint.Cap.BUTT
                        "round" -> currentPaint.strokeCap = Paint.Cap.ROUND
                        "square" -> currentPaint.strokeCap = Paint.Cap.SQUARE
                    }
                    when (currentState.lineJoin) {
                        "bevel" -> currentPaint.strokeJoin = Paint.Join.BEVEL
                        "miter" -> currentPaint.strokeJoin = Paint.Join.MITER
                        "round" -> currentPaint.strokeJoin = Paint.Join.ROUND
                    }
                    currentPaint.strokeWidth = (currentState.lineWidth * scale).toFloat()
                    currentPaint.strokeMiter = (currentState.miterLimit * scale).toFloat()
                    if (!currentState.currentTransform.isIdentity) {
                        drawingPath.reset()
                        drawingPath.addPath(currentPath)
                        drawingPath.transform(currentState.currentTransform)
                        currentCanvas.drawPath(drawingPath, currentPaint)
                        return@add
                    }
                    currentCanvas.drawPath(currentPath, currentPaint)
                }
                this.actions = actions.toList()
            }
            invalidate()
        }

        fun xtr_beginPath() {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentPath.reset()
                }
                this.actions = actions.toList()
            }
        }

        fun xtr_moveTo(value: V8Object) {
            XTRUtils.toPoint(value)?.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentPath.moveTo((it.x * scale).toFloat(), (it.y * scale).toFloat())
                    }
                    this.actions = actions.toList()
                }
            }
        }

        fun xtr_closePath() {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentPath.close()
                }
                this.actions = actions.toList()
            }
        }

        fun xtr_lineTo(value: V8Object) {
            XTRUtils.toPoint(value)?.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentPath.lineTo((it.x * scale).toFloat(), (it.y * scale).toFloat())
                    }
                    this.actions = actions.toList()
                }
            }
        }

        fun xtr_quadraticCurveTo(argCpPoint: V8Object, argXyPoint: V8Object) {
            val cpPoint = XTRUtils.toPoint(argCpPoint) ?: return
            val xyPoint = XTRUtils.toPoint(argXyPoint) ?: return
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentPath.quadTo((cpPoint.x * scale).toFloat(), (cpPoint.y * scale).toFloat(), (xyPoint.x * scale).toFloat(), (xyPoint.y * scale).toFloat())
                }
                this.actions = actions.toList()
            }
        }

        fun xtr_bezierCurveTo(argCp1Point: V8Object, argCp2Point: V8Object, argXyPoint: V8Object) {
            val cp1Point = XTRUtils.toPoint(argCp1Point) ?: return
            val cp2Point = XTRUtils.toPoint(argCp2Point) ?: return
            val xyPoint = XTRUtils.toPoint(argXyPoint) ?: return
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentPath.cubicTo((cp1Point.x * scale).toFloat(), (cp1Point.y * scale).toFloat(), (cp2Point.x * scale).toFloat(), (cp2Point.y * scale).toFloat(), (xyPoint.x * scale).toFloat(), (xyPoint.y * scale).toFloat())
                }
                this.actions = actions.toList()
            }
        }

        fun xtr_arc(argPoint: V8Object, r: Double, sAngle: Double, eAngle: Double, counterclockwise: Boolean) {
            val point = XTRUtils.toPoint(argPoint) ?: return
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    currentPath.addArc(
                            RectF(((point.x - r) * scale).toFloat(), ((point.y - r) * scale).toFloat(), ((point.x + r) * scale).toFloat(), ((point.y + r) * scale).toFloat()),
                            if (counterclockwise) ((eAngle / (2 * Math.PI)) * 360f).toFloat() else ((sAngle / (2 * Math.PI)) * 360f).toFloat(),
                            if (counterclockwise) ((sAngle / (2 * Math.PI)) * 360f).toFloat() - ((eAngle / (2 * Math.PI)) * 360f).toFloat() + 360f else ((eAngle / (2 * Math.PI)) * 360f).toFloat() - ((sAngle / (2 * Math.PI)) * 360f).toFloat()
                    )
                }
                this.actions = actions.toList()
            }
        }

        fun xtr_isPointInPath(value: V8Object): Boolean {
            XTRUtils.toPoint(value)?.let {
                actions.forEach { it.invoke(Canvas()) }
                val r = Region(0, 0, Int.MAX_VALUE, Int.MAX_VALUE)
                if (!currentState.currentTransform.isIdentity) {
                    drawingPath.reset()
                    drawingPath.addPath(currentPath)
                    drawingPath.transform(currentState.currentTransform)
                    r.setPath(drawingPath, Region(0, 0, Int.MAX_VALUE, Int.MAX_VALUE))
                }
                else {
                    r.setPath(currentPath, Region(0, 0, Int.MAX_VALUE, Int.MAX_VALUE))
                }
                return r.contains((it.x * scale).toInt(), (it.y * scale).toInt())
            }
            return false
        }

        fun xtr_postScale(value: V8Object) {
            XTRUtils.toPoint(value)?.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentState.currentTransform.postScale(it.x.toFloat(), it.y.toFloat())
                    }
                    this.actions = actions.toList()
                }
            }
        }

        fun xtr_postRotate(value: Double) {
            value.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentState.currentTransform.postRotate(((it / (2 * Math.PI)) * 360f).toFloat())
                    }
                    this.actions = actions.toList()
                }
            }
        }

        fun xtr_postTranslate(value: V8Object) {
            XTRUtils.toPoint(value)?.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentState.currentTransform.postTranslate((it.x * scale).toFloat(), (it.y * scale).toFloat())
                    }
                    this.actions = actions.toList()
                }

            }
        }

        fun xtr_postTransform(value: V8Object) {
            XTRUtils.toTransform(value)?.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentState.currentTransform.postConcat(it.toNativeMatrix())
                    }
                    this.actions = actions.toList()
                }

            }
        }

        fun xtr_setCanvasTransform(value: V8Object) {
            XTRUtils.toTransform(value)?.let {
                this.actions.toMutableList()?.let { actions ->
                    actions.add { _ ->
                        currentState.currentTransform = it.toNativeMatrix()
                    }
                    this.actions = actions.toList()
                }
            }
        }

        fun xtr_save() {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    val mutable = stateStack.toMutableList()
                    mutable.add(currentState)
                    stateStack = mutable.toList()
                    currentState = currentState.copy()
                }
                this.actions = actions.toList()
            }
        }

        fun xtr_restore() {
            this.actions.toMutableList()?.let { actions ->
                actions.add { _ ->
                    if (stateStack.isNotEmpty()) {
                        val mutable = stateStack.toMutableList()
                        val lastObject = mutable[mutable.size - 1]
                        mutable.removeAt(mutable.size - 1)
                        stateStack = mutable.toList()
                        currentState = lastObject
                    }
                }
                this.actions = actions.toList()
            }
        }

        fun xtr_clear() {
            actions = listOf()
            invalidate()
        }

        override fun drawContent(canvas: Canvas?) {
            super.drawContent(canvas)
            canvas?.takeIf { it.width > 0 && it.height > 0 }?.let { canvas ->
                stateStack = listOf()
                currentState = State()
                fakeState = State()
                currentPath.reset()
                drawingPath.reset()
                currentPaint.reset()
                actions.forEach { it.invoke(canvas) }
            }
        }

    }

}

File: ios/runtime/XTRCanvasView.m
//
//  XTRCanvasView.m
//  XTSample
//
//  Created by 崔明辉 on 2017/9/21.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTRCanvasView.h"
#import "XTRUtils.h"
#import "XTRContext.h"
#import <XT-Mem/XTMemoryManager.h>

@interface XTRCanvasState: NSObject<NSCopying>

@property (nonatomic, assign) CGFloat globalAlpha;
@property (nonatomic, strong) UIColor *fillStyle;
@property (nonatomic, strong) UIColor *strokeStyle;
@property (nonatomic, strong) NSString *lineCap;
@property (nonatomic, strong) NSString *lineJoin;
@property (nonatomic, assign) CGFloat lineWidth;
@property (nonatomic, assign) CGFloat miterLimit;
@property (nonatomic, assign) CGAffineTransform currentTransform;

@end

@implementation XTRCanvasState

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.globalAlpha = 1.0;
        self.lineWidth = 1.0;
        self.currentTransform = CGAffineTransformIdentity;
    }
    return self;
}

- (nonnull id)copyWithZone:(nullable NSZone *)zone {
    XTRCanvasState *newState = [XTRCanvasState allocWithZone:zone];
    newState.globalAlpha = self.globalAlpha;
    newState.fillStyle = self.fillStyle;
    newState.strokeStyle = self.strokeStyle;
    newState.lineCap = self.lineCap;
    newState.lineJoin = self.lineJoin;
    newState.lineWidth = self.lineWidth;
    newState.miterLimit = self.miterLimit;
    newState.currentTransform = self.currentTransform;
    return newState;
}

@end

@interface XTRCanvasView()

@property (nonatomic, strong) XTRCanvasState *currentState;
@property (nonatomic, copy) NSArray *stateStack;
@property (nonatomic, strong) UIBezierPath *currentPath;

@end

@implementation XTRCanvasView

+ (NSString *)name {
    return @"XTRCanvasView";
}

+ (NSString *)create:(JSValue *)frame scriptObject:(JSValue *)scriptObject {
    XTRCanvasView *view = [[XTRCanvasView alloc] initWithFrame:[frame toRect]];
    view.currentState = [XTRCanvasState new];
    view.backgroundColor = [UIColor clearColor];
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:view];
    [XTMemoryManager add:managedObject];
    view.context = [JSContext currentContext];
    view.objectUUID = managedObject.objectUUID;
    return managedObject.objectUUID;
}

+ (CGFloat)xtr_globalAlpha:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        return view.currentState.globalAlpha;
    }
    return 0.0;
}

+ (void)xtr_setGlobalAlpha:(CGFloat)globalAlpha objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.globalAlpha = globalAlpha;
    }
}

+ (NSDictionary *)xtr_fillStyle:(NSString *)objectRef{
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        return [JSValue fromColor:view.currentState.fillStyle];
    }
    return nil;
}

+ (void)xtr_setFillStyle:(JSValue *)fillStyle objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.fillStyle = [fillStyle toColor];
    }
}

+ (NSDictionary *)xtr_strokeStyle:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        return [JSValue fromColor:view.currentState.strokeStyle];
    }
    return nil;
}

+ (void)xtr_setStrokeStyle:(JSValue *)strokeStyle objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.strokeStyle = [strokeStyle toColor];
    }
}

+ (NSString *)xtr_lineCap:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        return view.currentState.lineCap;
    }
    return nil;
}

+ (void)xtr_setLineCap:(NSString *)lineCap objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.lineCap = lineCap;
    }
}

+ (NSString *)xtr_lineJoin:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        return view.currentState.lineJoin;
    }
    return nil;
}

+ (void)xtr_setLineJoin:(NSString *)lineJoin objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.lineJoin = lineJoin;
    }
}

+ (CGFloat)xtr_lineWidth:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        return view.currentState.lineWidth;
    }
    return 0.0;
}

+ (void)xtr_setLineWidth:(CGFloat)lineWidth objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.lineWidth = lineWidth;
    }
}

+ (CGFloat)xtr_miterLimit:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        return view.currentState.miterLimit;
    }
    return 0.0;
}

+ (void)xtr_setMiterLimit:(CGFloat)miterLimit objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.miterLimit = miterLimit;
    }
}

+ (void)xtr_rect:(JSValue *)rect objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentPath = [UIBezierPath bezierPathWithRect:[rect toRect]];
    }
}

+ (void)xtr_fillRect:(JSValue *)rect objectRef:(NSString *)objectRef {
    [self xtr_rect:rect objectRef:objectRef];
    [self xtr_fill:objectRef];
}

+ (void)xtr_strokeRect:(JSValue *)rect objectRef:(NSString *)objectRef {
    [self xtr_rect:rect objectRef:objectRef];
    [self xtr_stroke:objectRef];
}

+ (void)xtr_fill:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        if (view.currentPath != nil) {
            CAShapeLayer *layer = [CAShapeLayer layer];
            layer.transform = CATransform3DMakeAffineTransform(view.currentState.currentTransform);
            [layer setPath:[view.currentPath CGPath]];
            [layer setFillColor:[(view.currentState.fillStyle ?: [UIColor blackColor]) CGColor]];
            [layer setOpacity:view.currentState.globalAlpha];
            [layer setStrokeColor:[UIColor clearColor].CGColor];
            [view.layer addSublayer:layer];
        }
    }
}

+ (void)xtr_stroke:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        if (view.currentPath != nil) {
            CAShapeLayer *layer = [CAShapeLayer layer];
            layer.transform = CATransform3DMakeAffineTransform(view.currentState.currentTransform);
            [layer setPath:[view.currentPath CGPath]];
            [layer setStrokeColor:[(view.currentState.strokeStyle ?: [UIColor blackColor]) CGColor]];
            [layer setOpacity:view.currentState.globalAlpha];
            if ([view.currentState.lineCap isEqualToString:@"butt"]) {
                layer.lineCap = @"butt";
            }
            else if ([view.currentState.lineCap isEqualToString:@"round"]) {
                layer.lineCap = @"round";
            }
            else if ([view.currentState.lineCap isEqualToString:@"square"]) {
                layer.lineCap = @"suqare";
            }
            if ([view.currentState.lineJoin isEqualToString:@"bevel"]) {
                layer.lineJoin = @"bevel";
            }
            else if ([view.currentState.lineJoin isEqualToString:@"miter"]) {
                layer.lineJoin = @"miter";
            }
            else if ([view.currentState.lineJoin isEqualToString:@"round"]) {
                layer.lineJoin = @"round";
            }
            layer.lineWidth = view.currentState.lineWidth;
            layer.miterLimit = view.currentState.miterLimit;
            [layer setFillColor:[UIColor clearColor].CGColor];
            [view.layer addSublayer:layer];
        }
    }
}

+ (void)xtr_beginPath:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentPath = [[UIBezierPath alloc] init];
    }
}

+ (void)xtr_moveTo:(JSValue *)point objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        [view.currentPath moveToPoint:[point toPoint]];
    }
}

+ (void)xtr_closePath:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        [view.currentPath closePath];
    }
}

+ (void)xtr_lineTo:(JSValue *)point objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        [view.currentPath addLineToPoint:[point toPoint]];
    }
}

+ (void)xtr_quadraticCurveTo:(JSValue *)cpPoint xyPoint:(JSValue *)xyPoint objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        [view.currentPath addQuadCurveToPoint:[xyPoint toPoint] controlPoint:[cpPoint toPoint]];
    }
}

+ (void)xtr_bezierCurveTo:(JSValue *)cp1Point cp2Point:(JSValue *)cp2Point xyPoint:(JSValue *)xyPoint objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        [view.currentPath addCurveToPoint:[xyPoint toPoint]
                            controlPoint1:[cp1Point toPoint]
                            controlPoint2:[cp2Point toPoint]];
    }
}

+ (void)xtr_arc:(JSValue *)point r:(JSValue *)r sAngle:(JSValue *)sAngle eAngle:(JSValue *)eAngle counterclockwise:(JSValue *)counterclockwise objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        [view.currentPath addArcWithCenter:[point toPoint]
                                    radius:r.toDouble
                                startAngle:sAngle.toDouble
                                  endAngle:eAngle.toDouble
                                 clockwise:!counterclockwise.toBool];
    }
}

+ (BOOL)xtr_isPointInPath:(JSValue *)point objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        UIBezierPath *currentPath = view.currentPath;
        if (!CGAffineTransformIsIdentity(view.currentState.currentTransform)) {
            currentPath = [UIBezierPath bezierPath];
            [currentPath appendPath:view.currentPath];
            [currentPath applyTransform:view.currentState.currentTransform];
        }
        return [currentPath containsPoint:[point toPoint]];
    }
    return NO;
}

+ (void)xtr_postScale:(JSValue *)point objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.currentTransform = CGAffineTransformScale(view.currentState.currentTransform, [point toPoint].x, [point toPoint].y);
    }
}

+ (void)xtr_postRotate:(JSValue *)angle objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.currentTransform = CGAffineTransformRotate(view.currentState.currentTransform, angle.toDouble);
    }
}

+ (void)xtr_postTranslate:(JSValue *)point objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.currentTransform = CGAffineTransformTranslate(view.currentState.currentTransform, [point toPoint].x, [point toPoint].y);
    }
}

+ (void)xtr_postTransform:(JSValue *)transform objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        view.currentState.currentTransform = CGAffineTransformConcat(view.currentState.currentTransform, [transform toTransform]);
    }
}

+ (void)xtr_setCanvasTransform:(JSValue *)transform objectRef:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
    }
}

- (void)xtr_setCanvasTransform:(JSValue *)transform {
    self.currentState.currentTransform = [transform toTransform];
}

+ (void)xtr_save:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        NSMutableArray *stateStack = [(view.stateStack ?: @[]) mutableCopy];
        [stateStack addObject:view.currentState];
        view.stateStack = stateStack;
        view.currentState = [view.currentState copy];
    }
}

+ (void)xtr_restore:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        if ([view.stateStack count] > 0) {
            view.currentState = [view.stateStack lastObject];
            NSMutableArray *stateStack = [(view.stateStack ?: @[]) mutableCopy];
            [stateStack removeLastObject];
            view.stateStack = stateStack;
        }
    }
}

+ (void)xtr_clear:(NSString *)objectRef {
    XTRCanvasView *view = [XTMemoryManager find:objectRef];
    if ([view isKindOfClass:[XTRCanvasView class]]) {
        [view.layer.sublayers makeObjectsPerformSelector:@selector(removeFromSuperlayer)];
        view.stateStack = @[];
        view.currentState = [XTRCanvasState new];
        view.currentPath = nil;
    }
}

@end
